src/model.rs

```diff
@@ -787,88 +792,79 @@ impl Collections {
+                    // the number of vj checked should be limited
+                    other_block_id_vj
+                        .iter()
+                        .for_each(|(other_vj_idx, other_vj)| {
+                            if can_chain_without_overlap(vj, other_vj) {
+                                allowed_first_drop_off_vj.insert(*other_vj_idx);
+                                allowed_last_pick_up_vj.insert(vj_idx);
+                            } else if can_chain_without_overlap(other_vj, vj) {
+                                allowed_first_drop_off_vj.insert(vj_idx);
+                                allowed_last_pick_up_vj.insert(*other_vj_idx);
+                            }
+                        });
+                    other_block_id_vj.push((vj_idx, vj));
+                    map
+                },
+            );
+
+        let vj_idxs: Vec<Idx<VehicleJourney>> =
+            self.vehicle_journeys.iter().map(|(idx, _)| idx).collect();
         for vj_idx in &vj_idxs {
-            let mut vehicle_journey = self.vehicle_journeys.index_mut(*vj_idx);
-            vehicle_journey
-                .stop_times
-                .first_mut()
-                .unwrap()
-                .drop_off_type = 1;
-            vehicle_journey.stop_times.last_mut().unwrap().pickup_type = 1;
-        }
-        // Keep only vehicle journeys with a `block_id` (concerned with
-        // stay-in).
-        vj_idxs.retain(|vj_idx| self.vehicle_journeys[*vj_idx].block_id.is_some());
-        // Sort the vehicule journeys by block_id, and if equal, order them by
-        // departure time.
-        vj_idxs.sort_by(|vj1_idx, vj2_idx| {
-            let vj1 = &self.vehicle_journeys[*vj1_idx];
-            let vj2 = &self.vehicle_journeys[*vj2_idx];
-            match vj1.block_id.cmp(&vj2.block_id) {
-                Ordering::Equal => vj1
-                    .stop_times
-                    .first()
-                    .unwrap()
-                    .departure_time
-                    .cmp(&vj2.stop_times.first().unwrap().departure_time),
-                ordering => ordering,
-            }
-        });
-        for (prev_vj_idx, next_vj_idx) in vj_idxs.iter().zip(vj_idxs.iter().skip(1)) {
-            let prev_vj = &self.vehicle_journeys[*prev_vj_idx];
-            let next_vj = &self.vehicle_journeys[*next_vj_idx];
-            if prev_vj.block_id != next_vj.block_id {
-                // we can discard when two vehicle journeys are not in a stay-in
-                // situation.
-                continue;
-            }
-            let last_stop = &prev_vj.stop_times.last().unwrap();
-            let first_stop = &next_vj.stop_times.first().unwrap();
-            if last_stop.stop_point_idx == first_stop.stop_point_idx {
-                // We can discard when the stop points are identicals (see
-                // Example 1 above).
-                continue;
+            let mut vj = self.vehicle_journeys.index_mut(*vj_idx);
```

Comment:
It's uglier but for perf I think you should get that only if you really mutate it.

Reply:
it's just an access to en element by its position in a vector no?

Reply:
`RefMut` does a copy of `Idx` and a ref to collection (cheap) and of the `old_id`: https://docs.rs/typed_index_collection/1.2.1/src/typed_index_collection/collection.rs.html#1002-1006
And the `drop()` does some check: https://docs.rs/typed_index_collection/1.2.1/src/typed_index_collection/collection.rs.html#1018-1031

Looks like the interface on mutating CollectionWithId is not super straightforward (especially knowing that it's not zero-cost or minimal-cost).

Reply:
the only 'expensive' stuff here is the string and the cost should be quite limited, I think modifying code to bypass this is really premature optimisation